utils/s2n_random.c

```diff
@@ -180,32 +180,326 @@ S2N_RESULT s2n_get_urandom_data(struct s2n_blob *blob)
+     *      }
+     *  }
+     *
+     * On our number line this can be visualized as:
+     *
+     *  x =         0  1  2  3  4  5  6  7
+     *              +--+--+--+--+--+--+--+
+     *              \_____/  \_____/
+     *  x % 3 =     0  1  2  0  1  2
+     *
+     * if x comes up 6 or 7, we try again. That produces UINT3_MAX / s
+     * "ranges" of legitimate values. In this case two 0s, two 1s, two 2s.
+     *
+     * With our code, we checked if x was between 0 and 5, because that's
+     * easiest, but any contiguous window of 6 numbers would have done.
+     * For example:
+     *
+     *  x =         0  1  2  3  4  5  6  7
+     *              +--+--+--+--+--+--+--+
+     *                 \_____/  \_____/
+     *  x % 3 =        1  2  0  1  2  0
+     *
+     *  x =         0  1  2  3  4  5  6  7
+     *              +--+--+--+--+--+--+--+
+     *                    \_____/  \_____/
+     *  x % 3 =           2  0  1  2  0  1
+     *
+     * There's a general principle at play here. Any contiguous range of
+     * (n * s) numbers will contain exactly n values where x % s is 0, n values
+     * where x % s is 1, and so on, up to n values where x % s is (s - 1).
+     * This is important later, so really convince yourself of this.
+     *
+     * This algorithm works correctly but is expensive. There's at least two
+     * % operations per call and maybe more, and those operations are among the
+     * slowest a CPU can be asked to perform.
+     *
+     * To avoid this, we use Lemire's algorithm which cleverly replaces these
+     * modulus operations with bit-shifts. Here's how it works.
+     *
+     * First, we generate a random value as before:
+     *
+     *  uint3_t x = random3();
+     *
+     * which can be visualized on a number line:
+     *
+     *  x =        0  1  2  3  4  5  6  7
+     *             +--+--+--+--+--+--+--+
+     *
+     * We then multiply x by s. Recall that s is the size of the range we want,
+     * i.e. to pick a number in the set 0, 1, 2 then s is 3.
+     *
+     *  uint6_t m = x * s;
+     *
+     * Note that m is twice the bit-width of x. It's a 6-bit int, enough to represent
+     * the numbers 0-63. Since in our case s is 3, let's expand out all of the
+     * possibilities for m with another number line.
+     *
+     *  x =        0  1  2  3  4  5  6  7
+     *             +--+--+--+--+--+--+--+
+     *             .  .  .  .  ..  .. .. .......
+     *             .  .   .  ..  ... ..  ...    .......
+     *             .   .   .    ...  ....     .... ..   .......
+     *             .     .   .     ...   .....    ....  ....    .......
+     *             .      .    . .    ...     .....   .....  ....      ......
+     *             .      .        .     ....      ....    .....   .....     .....
+     *  m =        .        3        6        9        12       15       18       21
+     *             +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
+     *
```

Comment:
One last thing, I had a really hard time understanding what this chart was supposed to mean, here's the only alternative I've come up with: https://tiny.amazon.com/t2287hwa/paste. Maybe it's just me though 🤷‍♀️

Reply:
Your version is much better!

Reply:
I just did an algorithm review with a group, and we also sturggled with the chart. I'd like to commend maddeleine on this ASCII chart.